### PR TITLE
Feature / use span instead of reference wrapper for cache sequence

### PR DESCRIPTION
**Pull request by mgovers**
Part 1 of refactoring after optional ID

### CODE REVIEW

File: power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp [hunk 1]
```diff
@@ -435,7 +432,7 @@ class MainModelImpl<ExtraRetrievableTypes<ExtraRetrievableType...>, ComponentLis
                                      ComponentFlags const& to_store) const {
         // TODO: (jguo) this function could be encapsulated in UpdateCompIndependence in update.hpp
         return run_functor_with_all_types_return_array([this, scenario_idx, &update_data, &to_store]<typename CT>() {
-            if (!to_store[index_of_component<CT>]) {
+            if (!std::get<index_of_component<CT>>(to_store)) {
```
Comment (figueroa1395) [inline power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp:435]: Nitpick: Rename `to_store` in this function to something more expressive. Maybe `available_components`, or `components_to_store`?

File: power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp [hunk 2]
```diff
@@ -445,15 +442,7 @@ class MainModelImpl<ExtraRetrievableTypes<ExtraRetrievableType...>, ComponentLis
...
-        // TODO: (jguo) this function could be encapsulated in UpdateCompIndependence in update.hpp
-        return run_functor_with_all_types_return_array([this, &update_data, &to_store]<typename CT>() {
-            if (!to_store[index_of_component<CT>]) {
-                return std::vector<Idx2D>{};
-            }
-            auto const independence = check_components_independence<CT>(update_data);
-            validate_update_data_independence(independence);
-            return get_component_sequence<CT>(update_data, 0, independence);
-        });
+        return get_sequence_idx_map(update_data, 0, to_store);
```
Comment (figueroa1395) [inline power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp:445]: This is basically for permanent updates, right? Or for scenario 0 only. If so, shouldn't we edit as well the comment explaining the function?

**Review by nitbharambe — APPROVED** [review]

File: power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp [hunk 3]
```diff
@@ -720,28 +702,41 @@ class MainModelImpl<ExtraRetrievableTypes<ExtraRetrievableType...>, ComponentLis
...
+            return run_functor_with_all_types_return_array(
+                [&all_scenario_sequence, &current_scenario_sequence_cache, &is_independent]<typename CT>() {
+                    constexpr auto comp_idx = index_of_component<CT>;
+                    if (std::get<comp_idx>(is_independent)) {
+                        return std::span<Idx2D const>{std::get<comp_idx>(all_scenario_sequence)};
+                    }
+                    return std::span<Idx2D const>{std::get<comp_idx>(current_scenario_sequence_cache)};
+                });
+        };
+
```
Comment (figueroa1395) [inline power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp:726]: Curiousity: Why was this moved here from sub_batch_calculation? Just to reduce complexity on the later? Besides that, the "new" logic is clear.

**Review by figueroa1395 — COMMENTED** [review]
Looks good to me, but I prefer to not approve since we did the changes "together" in pair programming.

File: power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp (hunk 2 above)
Comment (mgovers) [reply power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp:445]: no, this is the entry point used when caching the entire batch after determining which component types are independent. the one below is for permanent updates.

will remove this as indeed it is very unclear as to what happens

File: power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp (hunk 3 above)
Comment (mgovers) [reply power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp:726]: > Just to reduce complexity on the latter?

yes and also to combine all the logic concerning which cache to use when into one single location.

**Review by figueroa1395 — APPROVED** [review]
I'm re-approving since @nitbharambe had approved already and the changes after are minor.

**Pull request merged**